hw/xbox/xid-sdl.c

```diff
@@ -512,14 +512,16 @@ static void usb_xbox_gamepad_realize(USBDevice *dev, Error **errp)
 #endif
 
 #ifdef FORCE_FEEDBACK
-    s->sdl_haptic = SDL_HapticOpenFromJoystick(s->sdl_gamepad);
+    s->sdl_haptic = SDL_HapticOpen(s->device_index);
     if (s->sdl_haptic == NULL) {
         fprintf(stderr, "Joystick doesn't support haptic\n");
     } else {
         if ((SDL_HapticQuery(s->sdl_haptic) & SDL_HAPTIC_LEFTRIGHT) == 0) {
             fprintf(stderr, "Joystick doesn't support necessary haptic effect\n");
             SDL_HapticClose(s->sdl_haptic);
             s->sdl_haptic = NULL;
+        } else {
+            s->sdl_haptic_effect_id = -1;
```

Comment:
`sdl_haptic_effect_id` is never assigned a value, so you played poker with your existing ram to see if we even enable haptic feedback. Assigning value here if device supports `SDL_HAPTIC_LEFTRIGHT`

Reply:
I don't think this is a good spot to set it though (rather hidden). I think it should happen independent of the haptic creation; somewhere where it's easily seen.

We should also consider creating the effect right here, so we have the feature query and effect creation together. The old method might cause a delay or lag when the effect is created during runtime, which is not ideal.

Reply:
Agreed on both counts. Since the effect is permanent it might just make more sense to store the effect on `s`'s struct instead of inferring it with the `id` anyway.

Reply:
`SDL_HapticEffect` should remain on the stack, because it's a stateless design: we just write to it and never read. We never have to do partial updates. It's a small structure, too.

The `id` should remain part of the state (we have to read it often), but be initialized once.